File: packages/go/canonicalpath/uri.go
package canonicalpath

import (
	"net/url"
	"strings"
)

// ParseFileURI unwraps an allowed file-like URI into a path string.
func ParseFileURI(uri string, opts Options) (string, error) {
	if strings.ContainsRune(uri, '\x00') {
		return "", newError(ErrNULByte, "URI contains NUL")
	}

	switch {
	case strings.HasPrefix(uri, "file://"):
		if !opts.URI.AllowFileURI {
			return "", newError(ErrUnsupportedURIScheme, "file URI is not allowed")
		}
		return parseHierarchicalURIPath(uri, "file://", opts)
	case strings.HasPrefix(uri, "vscode-file://"):
		if !opts.URI.AllowVSCodeFileURI {
			return "", newError(ErrUnsupportedURIScheme, "vscode-file URI is not allowed")
		}
		return parseHierarchicalURIPath(uri, "vscode-file://", opts)
	case hasURIScheme(uri):
		return "", newError(ErrUnsupportedURIScheme, "unsupported URI scheme")
	default:
		return uri, nil
	}
}

func parseHierarchicalURIPath(raw string, prefix string, opts Options) (string, error) {
	if rejectEncodedSlash(opts) && hasEncodedSeparator(raw) {
		return "", newError(ErrEncodedSeparator, "URI contains an encoded path separator")
	}

	rest := strings.TrimPrefix(raw, prefix)
	slash := strings.Index(rest, "/")
	if slash < 0 {
		return "", newError(ErrInvalidURI, "URI path is empty")
	}
	authority := rest[:slash]
	pathPart := rest[slash:]
	decoded, err := url.PathUnescape(pathPart)
	if err != nil {
		return "", newError(ErrInvalidPercentEncoding, "URI percent encoding is invalid")
	}
	decodedAuthority, err := url.PathUnescape(authority)
	if err != nil {
		return "", newError(ErrInvalidPercentEncoding, "URI authority percent encoding is invalid")
	}
	if decoded == "" {
		return "", newError(ErrInvalidURI, "URI path is empty")
	}
	if prefix == "file://" && decodedAuthority != "" && !strings.EqualFold(decodedAuthority, "localhost") {
		return "//" + decodedAuthority + decoded, nil
	}
	return decoded, nil
}

func rejectEncodedSlash(opts Options) bool {
	return opts.URI.RejectEncodedSlash == nil || *opts.URI.RejectEncodedSlash
}

func hasEncodedSeparator(value string) bool {
	for i := 0; i+2 < len(value); i++ {
		if value[i] != '%' {
			continue
		}
		hi, okHi := fromHex(value[i+1])
		lo, okLo := fromHex(value[i+2])
		if !okHi || !okLo {
			continue
		}
		decoded := hi<<4 | lo
		if decoded == '/' || decoded == '\\' {
			return true
		}
	}
	return false
}

func fromHex(b byte) (byte, bool) {
	switch {
	case b >= '0' && b <= '9':
		return b - '0', true
	case b >= 'a' && b <= 'f':
		return b - 'a' + 10, true
	case b >= 'A' && b <= 'F':
		return b - 'A' + 10, true
	default:
		return 0, false
	}
}

File: packages/ts/src/canonicalpath/uri.ts
import type { NormalizeOptions } from "./types.js";
import { pathError } from "./errors.js";
import { hasUriScheme } from "./internal.js";

export function parseFileUri(uri: string, options: NormalizeOptions = {}): string {
  if (uri.includes("\0")) throw pathError("ERR_NUL_BYTE", "URI contains NUL");

  if (uri.startsWith("file://")) {
    if (!options.uri?.allowFileUri) throw pathError("ERR_UNSUPPORTED_URI_SCHEME", "file URI is not allowed");
    return parseHierarchicalURIPath(uri, "file://", options);
  }
  if (uri.startsWith("vscode-file://")) {
    if (!options.uri?.allowVSCodeFileUri) throw pathError("ERR_UNSUPPORTED_URI_SCHEME", "vscode-file URI is not allowed");
    return parseHierarchicalURIPath(uri, "vscode-file://", options);
  }
  if (hasUriScheme(uri)) throw pathError("ERR_UNSUPPORTED_URI_SCHEME", "unsupported URI scheme");
  return uri;
}

function parseHierarchicalURIPath(raw: string, prefix: string, options: NormalizeOptions): string {
  if (options.uri?.rejectEncodedSlash !== false && hasEncodedSeparator(raw)) {
    throw pathError("ERR_ENCODED_SEPARATOR", "URI contains an encoded path separator");
  }

  const rest = raw.slice(prefix.length);
  const slash = rest.indexOf("/");
  if (slash < 0) throw pathError("ERR_INVALID_URI", "URI path is empty");

  const authority = rest.slice(0, slash);
  const pathPart = rest.slice(slash);
  try {
    const decoded = decodeURIComponent(pathPart);
    const decodedAuthority = decodeURIComponent(authority);
    if (decoded === "") throw pathError("ERR_INVALID_URI", "URI path is empty");
    if (prefix === "file://" && decodedAuthority !== "" && decodedAuthority.toLowerCase() !== "localhost") {
      return `//${decodedAuthority}${decoded}`;
    }
    return decoded;
  } catch (error) {
    if (error instanceof Error && error.name === "CanonicalPathError") throw error;
    throw pathError("ERR_INVALID_PERCENT_ENCODING", "URI percent encoding is invalid");
  }
}

function hasEncodedSeparator(value: string): boolean {
  return /%(2f|2F|5c|5C)/.test(value);
}
